### PR TITLE
rgw/sfs: added fields in sqlite buckets table

### PR DESCRIPTION
**Pull request by giubacc**
- Added the following columns in sqlite buckets table:

	- zone_group
	- quota
	
- Some code cleaning and optimization
- Updated test_rgw_sfs_sqlite_buckets.cc

Fixes: https://github.com/aquarist-labs/s3gw/issues/85
Signed-off-by: Giuseppe Baccini <giuseppe.baccini@suse.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/rgw/store/sfs/sfs_bucket.cc
Comment (jecluis) [inline src/rgw/store/sfs/sfs_bucket.cc:46]: any particular reason why we're doing a copy here?

File: src/rgw/store/sfs/sqlite/buckets/bucket_definitions.h
Comment (jecluis) [inline src/rgw/store/sfs/sqlite/buckets/bucket_definitions.h:26]: for future reference, the difference between `specialise` and `specialize` is just a matter of whether it's UK or not.

File: src/rgw/store/sfs/user.cc
Comment (jecluis) [inline src/rgw/store/sfs/user.cc:180]: please drop this to `0` instead of `1`.

**Review by jecluis — CHANGES_REQUESTED** [review]

File: src/rgw/rgw_sal_sfs.h
Comment (giubacc) [reply src/rgw/rgw_sal_sfs.h:351]: I admit I don't know much about swift; I added it just for similarity with other BEs behavior.
We can drop it anyway if you think we should.

File: src/rgw/rgw_sal_sfs.h
Comment (giubacc) [reply src/rgw/rgw_sal_sfs.h:357]: Actually I didn't run the python test checking the response, I will do it (we will have to run this test automatically with CI at some point).
Btw, I think this shouldn't be broken cause there is then this stmt at line 379:
```
info.objv_tracker = objv_tracker;
```

File: src/rgw/store/sfs/CHANGELOG.md
Comment (giubacc) [reply src/rgw/store/sfs/CHANGELOG.md:10]: will do :+1:

File: src/rgw/store/sfs/sfs_bucket.cc
Comment (giubacc) [reply src/rgw/store/sfs/sfs_bucket.cc:46]: Can you explain better what you mean?
This is just a style patch not directly related to this PR; I used this PR to do this and not opened a dedicated one because it is (is it :smirk: ?)  reasonably safe.

File: src/rgw/store/sfs/sqlite/buckets/bucket_definitions.h
Comment (giubacc) [reply src/rgw/store/sfs/sqlite/buckets/bucket_definitions.h:26]: ok

File: src/rgw/store/sfs/user.cc
Comment (giubacc) [reply src/rgw/store/sfs/user.cc:180]: will do :+1:

File: src/rgw/rgw_sal_sfs.h
Comment (giubacc) [reply src/rgw/rgw_sal_sfs.h:357]: tested, this is working as expected.

File: src/rgw/rgw_sal_sfs.h
Comment (giubacc) [reply src/rgw/rgw_sal_sfs.h]: Removed swift related fields.

File: src/rgw/store/sfs/sfs_bucket.cc
Comment (0xavi0) [reply src/rgw/store/sfs/sfs_bucket.cc:46]: I think this is related to a discussion we had in a previous PR.
In this particular case, I think the compiler will be doing copy elision and moving the structure (returning its local variable), so no copy should be involved as this is used in a constructor as destination memory is not initialised before the call.

Comment (giubacc) [reply src/rgw/store/sfs/sfs_bucket.cc:46]: Yes I remember that.
An ever more efficient operation here would be to remove the 3rd param from the `SFSBucket` constructor and call `to_rgw_bucket_info` over the bucket object (so no copy of info would exist at all).
But considering that we are talking about a `create` bucket operation I don't think we have a special need to super-optimize this call.

File: src/rgw/rgw_sal_sfs.h
Comment (0xavi0) [inline src/rgw/rgw_sal_sfs.h:355]: Regarding this... I've created a bug (https://github.com/aquarist-labs/s3gw/issues/103) because we'd need to implement the logic of version tracking in metadata.

It works here because we're really doing nothing aside from storing the initial value.

I also suspect we should assign the value coming from `info` to consume version tracking instead of initialising every time, but that will possibly arise when we implement the version tracking logic. (and also I'm not 100% sure if we need it for buckets here or initialising when creating the bucket is fine)

Have you checked what's coming in the `info` value? I guess it's already an initialised value. 

So if assigning what's coming from `info` is breaking this PR I'm happy to leave it as it is and change it when logic is implemented (if needed).. if it's not breaking it I would store whatever is received in the `info` variable.

**Review by 0xavi0 — COMMENTED** [review]
Just minor comment about version tracking that might worth checking. 
Happy to approve when that's clarified

File: src/rgw/store/sfs/sfs_bucket.cc
Comment (0xavi0) [reply src/rgw/store/sfs/sfs_bucket.cc:46]: yes, removing the 3rd parameter and doing it in the constructor sounds good to me, but I would not consider it a blocker for approving this PR.
Anyway.. it is something that we do every time we construct the `SFSBucket` when we pass the `sfs::BucketRef`, right? 
Makes sense doing it in that particular constructor.

Comment (giubacc) [reply src/rgw/store/sfs/sfs_bucket.cc:46]: yes, it is something we do at every `SFSBucket` construction.
ok, I will modify that (removing 3rd param and do it insisde ctor), better do it now.

Comment (giubacc) [reply src/rgw/store/sfs/sfs_bucket.cc:46]: mhmmh, no, this is not safe as I thought, doing this change in the context of this PR is potentially less safe than doing it in a dedicated one.
The problem here is that we have to rely on transformations from `rgw::sal::sfs::Bucket` to get a `RGWBucketInfo`.
I think something on `rgw::sal::sfs::Bucket` is missing; IMO the best choice for a future refactoring is to consider the idea to drop entirely `rgw::sal::sfs::Bucket` and rely only on `SFSBucket`.

File: src/rgw/rgw_sal_sfs.h
Comment (giubacc) [reply src/rgw/rgw_sal_sfs.h:355]: The value coming from info is `empty`. 
The value must be computed initially in `create` function.
I think a value here was needed for the UI to work.

Comment (jecluis) [reply src/rgw/rgw_sal_sfs.h:355]: what if it's a versioned bucket? Is the value still empty?

Comment (giubacc) [reply src/rgw/rgw_sal_sfs.h:355]: The computation of a version id is always done in `create` function.
This has not been introduced with this PR btw, if we want to do changes to this behavior I suggest do it on another dedicated one.

Comment (jecluis) [reply src/rgw/rgw_sal_sfs.h:355]: That's not what I meant. The bucket versioning is enabled on a bucket by the user. I would assume that this `objv_tracker` only makes sense when the bucket versioning is enabled. While you are getting an empty info in this case, might not mean it's empty if bucket versioning is enabled. Hence my question.

Comment (giubacc) [reply src/rgw/rgw_sal_sfs.h:355]: can we move these checks for another issue?

Comment (jecluis) [reply src/rgw/rgw_sal_sfs.h:355]: I don't think we can. We need to make sure this is not causing a problem.

File: src/rgw/store/sfs/sfs_bucket.cc
Comment (jecluis) [reply src/rgw/store/sfs/sfs_bucket.cc:46]: How do you suggest we track bucket-specific state with just `SFSBucket`?

File: src/rgw/rgw_sal_sfs.h
Comment (giubacc) [reply src/rgw/rgw_sal_sfs.h:355]: Keep in mind that if this is causing problems it already does since this [PR](https://github.com/aquarist-labs/ceph/pull/27) has been merged.
but ok, no problem for me, let's delve into the problem.

Comment (jecluis) [reply src/rgw/rgw_sal_sfs.h:355]: I don't know if what has been merged is causing a problem, but the worry is that this specific change, on this specific PR, may be introducing a problem. Keep in mind that this particular change is essentially squashing whatever `objv_tracker` is coming from the `info` function parameter with an empty one (the one you are creating) somewhere down the line.

I don't think it's unreasonable to expect validation on this change before merging.

File: src/rgw/store/sfs/sfs_bucket.cc
Comment (giubacc) [reply src/rgw/store/sfs/sfs_bucket.cc:46]: I don't know; I was just wondering if could be an idea for the future try to simplify this aspect.
But perhaps there are constraint that are preventing this.
Having just one subclass for sfs bucket is something desirable because it avoid completely to transfer information from one type to another.
But if we cannot have this, "coraggio" (courage :).

File: src/rgw/rgw_sal_sfs.h
Comment (giubacc) [reply src/rgw/rgw_sal_sfs.h:355]: understand and I agree; better a check more than one check less. ;)

Comment (giubacc) [reply src/rgw/rgw_sal_sfs.h]: removed bucket version related code and columns.

File: src/rgw/store/sfs/sfs_bucket.cc
Comment (giubacc) [reply src/rgw/store/sfs/sfs_bucket.cc:46]: I think we can close this thread, if anyone thinks that there is a better approach to model bucket objects in `sfs` is always welcome to propose an idea.
From my side, if I will have time, I will try to hack something.

**Review by 0xavi0 — APPROVED** [review]
lgtm

File: src/rgw/rgw_sal_sfs.h
Comment (jecluis) [inline src/rgw/rgw_sal_sfs.h:369]: This might have been mistakenly removed. I don't think we want to drop this line.

**Review by jecluis — CHANGES_REQUESTED** [review]

File: src/rgw/rgw_sal_sfs.h
Comment (giubacc) [reply src/rgw/rgw_sal_sfs.h:369]: ugh, nice catch.

Comment (giubacc) [reply src/rgw/rgw_sal_sfs.h:369]: this would have been prevented if python tests would have run on CI on push; this is something we will have to add to CI :smirk:

Comment (giubacc) [reply src/rgw/rgw_sal_sfs.h:369]: @jecluis checked better this, and that remove is intended.
now that code has been substituted with:

`info.placement_rule = db_binfo.binfo.placement_rule;`

I have run the bucket python tests and they must be fixed in agreement with the removal of version attributes.
But this is another story.

**Review by jecluis — APPROVED** [review]

**Pull request merged**
